tools/codegen/gen_functionalization_type.py

```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
+                assert g.view_inplace.has_composite_implicit_autograd_kernel
+                view_str.append(emit_registration_helper(g.view_inplace))
+            return view_str
+
+        # functionalization needs to register kernels for view + view_inplace ops
+        if funcTarget == FunctionalizationTarget.FUNCTIONALIZATION:
+            view_str = [emit_registration_helper(g.view)]
+            if g.view_inplace is not None:
+                assert g.view_inplace.is_view_op
+                view_str.append(emit_registration_helper(g.view_inplace))
+            return view_str
+
+        # while "add back view" kernels need to register kernels for view_copy ops
+        assert funcTarget == FunctionalizationTarget.ADD_BACK_VIEWS
+        assert g.view_copy is not None
+        return [emit_registration_helper(g.view_copy)]
     else:
         f = g
-        return [emit_registration_helper(f, is_view=False)]
+        assert not f.is_view_op
+        return [emit_registration_helper(f)]
 
 def gen_functionalization_definition(
     selector: SelectiveBuilder,
     g: Union[NativeFunction, NativeFunctionsViewGroup],
-    functional_op: Optional[NativeFunction]
+    functional_op: Optional[NativeFunction],
+    funcTarget: FunctionalizationTarget,
 ) -> List[str]:
-    if not needs_functionalization(selector, g):
+    if not needs_functionalization(selector, g, Target.ANONYMOUS_DEFINITION, funcTarget):
         return []
-    if isinstance(g, NativeFunctionsViewGroup):
-        view_defs = []
-        if not g.view.has_composite_implicit_autograd_kernel:
-            # For now I'm just asserting this - later, the view functionalization kernels
-            # should be updated to redispatch to view_copy().
-            assert g.view_copy is not None
-            view_defs.append(emit_view_functionalization_body(g.view, g.view))
-        if g.view_inplace is not None and not g.view_inplace.has_composite_implicit_autograd_kernel:
-            view_defs.append(emit_view_functionalization_body(g.view_inplace, g.view))
-        return view_defs
-    else:
-        f = g
-        assert modifies_arguments(f)
-        return [emit_inplace_functionalization_body(f, functional_op)]
+    if funcTarget == FunctionalizationTarget.FUNCTIONALIZATION:
+        if isinstance(g, NativeFunctionsViewGroup):
+            # Case 1: emit view -> view_copy kernels for the functionalization pass
+            view_defs = []
+            if not g.view.has_composite_implicit_autograd_kernel:
+                # invariant: NativeFunctionsViewGroup's always have a view_copy operator
+                # if the view is not composite (implicit autograd)
+                assert g.view_copy is not None
+                view_defs.append(emit_view_functionalization_body(g.view, g.view_copy))
+            if g.view_inplace is not None and not g.view_inplace.has_composite_implicit_autograd_kernel:
```

Comment:
is it ever the case that the view inplace has composite implicit autograd, but the view doesn't (or vice versa?)

Reply:
(commented above - as of today, nope)